torch/lib/THD/base/data_channels/DataChannelGloo.hpp

```diff
@@ -0,0 +1,319 @@
+      context,
+      std::make_shared<::gloo::AllgatherRing<T>>(
+        *context,
+        {reinterpret_cast<T*>(input_buffer->data())},
+        reinterpret_cast<T*>(output_buffer->data()),
+        count),
+      input_buffer,
+      output_buffer
+    );
+  }
+};
+
+template<typename T>
+struct algorithm_spec<DataOperation::ALL_REDUCE, T> {
+  static GlooCache::key_type key(
+    THDGroup group_id, std::size_t input_bytes,
+    std::size_t count /* unused */, THDReduceOp op
+  ) {
+    return std::make_tuple(DataOperation::ALL_REDUCE, group_id,
+                           input_bytes, input_bytes, op);
+  }
+
+  static GlooCache::value_type create(
+    const DataChannel::Group& group,
+    std::size_t input_bytes, std::size_t count, THDReduceOp op
+  ) {
+    auto& cache = GlooCache::get();
+    auto context = cache.createContext(group);
+    auto input_buffer = cache.createBuffer(input_bytes);
+    return std::make_tuple(
+      context,
+      std::make_shared<::gloo::AllreduceRing<T>>(
+        *context, {reinterpret_cast<T*>(input_buffer->data())},
+        count, THDToGlooReduceOp<T>(op)),
+      input_buffer,
+      input_buffer // we get the result in same buffer
+    );
+  }
+};
+
+template<typename T>
+struct algorithm_spec<DataOperation::BROADCAST, T> {
+  static GlooCache::key_type key(
+    THDGroup group_id, std::size_t input_bytes,
+    std::size_t count, rank_type src_rank // unused
+  ) {
+    return std::make_tuple(DataOperation::BROADCAST, group_id,
+                           input_bytes, input_bytes, 0);
+  }
+
+  static GlooCache::value_type create(
+    const DataChannel::Group& group,
+    std::size_t input_bytes, std::size_t count, rank_type src_rank
+  ) {
+    auto& cache = GlooCache::get();
+    auto context = cache.createContext(group);
+    auto input_buffer = cache.createBuffer(input_bytes);
+    return std::make_tuple(
+      context,
+      std::make_shared<::gloo::BroadcastOneToAll<T>>(
+        *context, {reinterpret_cast<T*>(input_buffer->data()}, count, src_rank
+      ),
+      input_buffer,
+      input_buffer // we get the result in same buffer
+    );
+  }
+};
+
+template<typename T> // unused
+struct algorithm_spec<DataOperation::BARRIER, T> {
+  static GlooCache::key_type key(THDGroup group_id) {
+    return std::make_tuple(DataOperation::BARRIER, group_id, 0, 0, 0);
+  }
+
+  static GlooCache::value_type create(const DataChannel::Group& group) {
+    auto context = GlooCache::get().createContext(group);
+    return std::make_tuple(
+      context,
+      std::make_shared<::gloo::BarrierAllToAll>(*context),
+      nullptr,
+      nullptr
+    );
+  }
+};
+
+
+struct GlooCache {
+  using key_type = std::tuple<
+    DataOperation, THDGroup, std::size_t, std::size_t, THDReduceOp
+  >;
+  using value_type = std::tuple<
+    std::shared_ptr<context_type>,
+    std::shared_ptr<algorithm_type>,
+    std::shared_ptr<buffer_type>,
+    std::shared_ptr<buffer_type>
+  >;
+
+  using ret_type = std::tuple<
```

Comment:
I don't think they're even necessary. If you change the vectors to `shared_ptr<real[]>`, and remove the pointer to the context, you could just use a single type (algorithms have a `shared_ptr` to the context, so you don't need to keep it alive yourself).